INT-4257: Introduce ErrorMessagePublishingRecoveryCallback

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4257

To make a target `ErrorMessage` customizable introduce
`ErrorMessagePublishingRecoveryCallback` and
`ErrorMessageStrategy` to inject

**Cherry-pick to 4.3.x**

artembilan committed:
INT-4257: Introduce ErrMessagePubRecoveryCallback

JIRA: https://jira.spring.io/browse/INT-4257

To make a target `ErrorMessage` customizable introduce
`ErrorMessagePublishingRecoveryCallback` and
`ErrorMessageStrategy` to inject

**Cherry-pick to 4.3.x**

artembilan committed:
Fix `ErrorMessagePublishingRecoveryCallback` generic type for
`RequestHandlerRetryAdvice` compatibility

garyrussell committed:
Polishing

- Support publishing when no retry context is available
- Use a constant for the message context key

garyrussell commented:
Suggested polish here: https://github.com/garyrussell/spring-integration/commit/f1929f590d08401c4ddb7ec381b1f12e5b7f440a

garyrussell committed:
Add EnhancedErrorMessage

- contains input message at time of error message generation
- make the default RecovererErrorMessageStrategy public and extensible

garyrussell commented:
More polish - add `EnhancedErrorMessage` https://github.com/garyrussell/spring-integration/commit/63cb9b84945a04bca0c1eba004aa01452ea792b1

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/handler/advice/ErrorMessageSendingRecoverer.java:
```diff
@@ -40,46 +32,37 @@
-		this.messagingTemplate.setDefaultDestination(channel);
+		setRecoveryChannel(channel);
+		setErrorMessageStrategy(new RecovererErrorMessageStrategy());
 	}
```
Needs a 3rd ctor so we can pass in a strategy.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/support/ErrorMessagePublishingRecoveryCallback.java:
```diff
@@ -0,0 +1,134 @@
+ *
+ * @since 4.3.9
+ */
+public class ErrorMessagePublishingRecoveryCallback implements RecoveryCallback<Object>, BeanFactoryAware {
```
`s/Object/Void/`

garyrussell commented:
Final Polishing: https://github.com/garyrussell/spring-integration/commit/ffff4c229ea19cc810a5388c44ecb3191ff19243

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/message/EnhancedErrorMessage.java:
```diff
@@ -0,0 +1,61 @@
+ * @since 4.3.9
+ *
+ */
+public class EnhancedErrorMessage extends ErrorMessage {
```
I don't like it either; I started with `Composite` but it's not a composite.

IAEM is ok, I guess, but I'm not thrilled with that either.

artembilan replied:
OK. Since we can't come up with something what sounds better and still provides the proper context, I won't mind to live with this name.

@mbogoevici , maybe you have some ideas? Thanks

mbogoevici replied:
One possible idea (for better or worse) - `RequestAwareErrorMessage` and rename `inputMessage` to `requestMessage` ?

artembilan committed:
Fix `RetryAdviceParserTests` for new `RecoveryCallback` architecture

garyrussell committed:
Decouple ErrorMessagePublisher from Retry

garyrussell commented:
Decouple `ErrorMessagePublisher` from `spring-retry`... https://github.com/garyrussell/spring-integration/commit/2667b739b007a785dea22eda7271d1200579cb3b

artembilan commented:
> Decouple `ErrorMessagePublisher`

OK. Will come back to you with some polishing soon.

Thank you for looking forward to this!
I really is distracted with other stuff to think more here 😢

artembilan committed:
Refactoring for better purpose reflection and JavaDocs

artembilan commented:
Pushed some further refactoring and polishing.

mbogoevici commented on spring-integration-core/src/main/java/org/springframework/integration/endpoint/MessageProducerSupport.java:
```diff
@@ -173,12 +173,22 @@ protected void sendMessage(Message<?> message) {
 			MessageChannel errorChannel = getErrorChannel();
 			if (errorChannel != null) {
-				this.messagingTemplate.send(errorChannel, new ErrorMessage(e));
+				this.messagingTemplate.send(errorChannel, buildErrorMessage(e));
```
Wouldn't it be preferable to delegate this to an injected `ErrorMessageStrategy` rather than invoking the overridable method below ?

artembilan replied:
The question relates to the change here: https://github.com/spring-projects/spring-integration-kafka/pull/163/files#diff-b49c842da794eb3622242c8acb03e1c1R290

I don't know the story yet, but let @garyrussell to answer!

garyrussell replied:
Yes, that is my plan; I meant to add that change in my latest polishing commit.

artembilan replied:
I got it! Let me fix this concern 😄

garyrussell replied:
Oops - here's my version: https://github.com/garyrussell/spring-integration/commit/6a4ed7604a6570e3981779d0381df533e5961254

artembilan replied:
Ha-ha! OK. I give up on this stuff.
Is there some work which you really can't do and I am still would be useful?  😉

garyrussell replied:
Sorry - I have sort-of hijacked your PR ☹️

artembilan replied:
Well, that doesn't look so bad (hijacking) because my mind doesn't go somehow so far 😢

garyrussell committed:
Use `ErrorMessageStrategy` in message producers.

Add JavaDocs to the `DefaultErrorMessageStrategy`

artembilan commented:
Applied your changes Gary.
Awesome work as usual!
I have just added JavaDocs to that `DefaultErrorMessageStrategy`.

My idea meanwhile was to use that new `ErrorMessagePublisher` as an internal instance for the provided `ErrorMessageStrategy` and don't expose the `DefaultErrorMessageStrategy` to the top level class.
But your approach is fine as well.
Anyway we have to keep the behavior like re-throw exception when the `errorChannel` isn't provided.

@mbogoevici , I think we can live with that `EnhancedErrorMessage` class until `5.0`. The JIRA https://jira.spring.io/browse/SPR-15459 has been just fixed and `ErrorMessage` will have `originalMessage` property since there.

I'm adding a deprecation note there though...

garyrussell commented:
Perhaps for 5.0 only, we should consider reworking the MPEH to use this strategy from pollers and executors too.

artembilan commented:
> we should consider reworking the MPEH to use this strategy

Let's do that in separate PRs, @garyrussell !

garyrussell commented:
agreed

artembilan committed:
Deprecate EnhancedErrorMessage

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/support/ErrorMessageUtils.java:
```diff
@@ -0,0 +1,72 @@
+	}
+
+	@SuppressWarnings("serial")
+	private static class ErrorMessageAttributes extends AttributeAccessorSupport {
```
Do you mind if I make this public? I need it in S-I-K. Or, we could make it a top level class.

garyrussell replied:
Actually, no; forget that.

garyrussell commented:
Minor polish needed for S-I-K: https://github.com/garyrussell/spring-integration/commit/1797e95dc4aceff2256128d11044fde12899309e

artembilan commented:
The latest polishing is fine with me.
Let me know if there is need to be done with this PR yet!

garyrussell commented:
If you are happy; I will merge and cherry-pick, and we can move over to S-I-K 😄

artembilan commented:
Ship it!
We always can add subsequent commits if that 😄

garyrussell commented:
Merged as 28058884f4c9ec1c99e5d89f7360d08ce0d1e41e cherry-picked as f7adfc06286fc5171592839fd78dae81a030e248
